frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -4538,71 +4554,96 @@ const CompositeType* isNameOfField(Context* context,
+    if (auto ct2 = ft->toCompositeType()) {
+      // Recursion of record types is never legal.
+      if (ct2 == ct && ct->isRecordType()) return false;
 
-    if (!rf.isGeneric()) return true;
+      // Bail to avoid an infinite loop.
+      if (context->isQueryRunning(canDefaultInitCompositeQuery, {ct2})) {
+        return false;
+      }
+    }
 
-    // TODO: Do I still need to consider field genericity, here? I.E., if
-    // a field is marked 'GENERIC_WITH_DEFAULTS' is there more to do?
-    // If I can tell the thing is concrete from the ResolvedFields, then
-    // there's probably no need to recurse.
-    if (rf.isGenericWithDefaults()) {
-      for (int i = 0; i < rf.numFields(); i++) {
-        auto ft = rf.fieldType(i).type();
+    // Otherwise, consider the field.
+    if (!rf.fieldHasDefaultValue(i) &&
+        !isTypeWithDefaultValue(context, ft)) {
+      allFieldsInitializable = false;
+      break;
+    }
+  }
 
-        // TODO: Skipping avoids a recursive query but doesn't handle
-        // mutually recursive classes.
-        if (ft == t) continue;
+  const uast::AstNode* ast = nullptr;
+  if (auto& id = ct->id()) ast = parsing::idToAst(context, id);
 
-        if (!isTypeDefaultInitializable(context, ft)) return false;
-      }
+  auto tfs = tryResolveDefaultInit(context, ast, ct);
 
-      return true;
+  // If the user wrote a default initializer, check to see if it is valid.
+  // TODO: Check for 'compilerError' in the default initializer body.
+  // TODO: Handle 'WHERE_TBD'?
+  if (tfs && !tfs->isCompilerGenerated()) {
+    auto w = tfs->whereClauseResult();
+    switch (w) {
+      case TypedFnSignature::WHERE_FALSE: return false;
+      case TypedFnSignature::WHERE_NONE: return true;
+      case TypedFnSignature::WHERE_TRUE: return true;
+      case TypedFnSignature::WHERE_TBD: break;
     }
   }
 
-  return false;
+  return allFieldsInitializable;
 }
 
 static const bool&
-isTypeDefaultInitializableQuery(Context* context, const Type* t) {
-  QUERY_BEGIN(isTypeDefaultInitializableQuery, context, t);
-  bool ret = isTypeDefaultInitializableImpl(context, t);
+canDefaultInitCompositeQuery(Context* context, const CompositeType* ct) {
+  QUERY_BEGIN(canDefaultInitCompositeQuery, context, ct);
+  bool ret = canDefaultInitCompositeImpl(context, ct);
   return QUERY_END(ret);
 }
 
-bool isTypeDefaultInitializable(Context* context, const Type* t) {
-  return isTypeDefaultInitializableQuery(context, t);
+bool isTypeWithDefaultValue(Context* context, const Type* t) {
+  if (t == nullptr) return false;
+  if (t->isUnknownType() || t->isErroneousType()) return false;
+  if (t == CompositeType::getStringType(context)) return true;
+  if (t == CompositeType::getBytesType(context)) return true;
+  if (t == CompositeType::getRangeType(context)) return true;
+  if (t == CompositeType::getLocaleType(context)) return true;
+  if (t->isPrimitiveType()) return true;
+  if (getTypeGenericity(context, t) == Type::GENERIC) return false;
```

Comment:
I suggest swapping this line with the previous, since something like `integral` might be considered a primitive type, but can't be default initialized